File: ios-app/CalendarioCiclismo/Views/Components/CountryFlag.swift
import SwiftUI

/// Bandera de país rectangular (20×15 pt). Los SVG viven empaquetados en
/// `Assets.xcassets/Flags/<code>.imageset` con `preserves-vector-representation`
/// activado, así que iOS los renderiza nativamente como vectores a cualquier
/// escala — sin red, sin `WKWebView`, sin rasterización manual, sin parpadeo.
///
/// El código ISO se pasa en minúsculas, lo que permite banderas sub-nacionales
/// reales (es-ct, es-pv, gb-eng, gb-sct, gb-wls) ya incluidas en el set.
struct CountryFlag: View {
    let countryCode: String?
    /// Ancho de la bandera; el alto se deriva con la proporción 4:3 (20×15).
    /// Por defecto 20 pt, que es el tamaño usado en listas y cabeceras.
    var width: CGFloat = 20

    private var height: CGFloat { width * 15 / 20 }

    var body: some View {
        let code = countryCode?.lowercased() ?? ""
        if !code.isEmpty {
            Image("Flags/\(code)")
                .resizable()
                .interpolation(.high)
                .scaledToFill()
                .frame(width: width, height: height)
                .clipShape(RoundedRectangle(cornerRadius: 2))
                .accessibilityLabel(AccessibilityCountryNames.name(for: countryCode) ?? "")
        }
    }
}

File: android-app/app/src/main/java/app/calendariociclismo/android/ui/components/CountryFlag.kt
package app.calendariociclismo.android.ui.components

import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import coil3.compose.AsyncImage
import coil3.request.ImageRequest

/**
 * Bandera de país rectangular (proporción 4:3, default 20×15 dp). Los SVG
 * viven empaquetados en `app/src/main/assets/flags/<code>.svg` — mismo set
 * que usa iOS (lipis/flag-icons v7.2.3, 4x3). Se cargan vía
 * `file:///android_asset/...` y los decodifica Coil con su `SvgDecoder` ya
 * registrado en `CalendarioCiclismoApp.newImageLoader`.
 *
 * Al ir por disco local no hay red, no hay parpadeo al cambiar de día ni al
 * entrar en otras vistas, y el render funciona offline desde el primer arranque.
 *
 * El código se pasa en minúsculas, lo que permite banderas sub-nacionales
 * (es-ct, es-pv, gb-eng, gb-sct, gb-wls) ya incluidas en el set.
 *
 * `height` permite ajustar el tamaño en contextos densos (p.ej. orden de
 * salida usa 11dp para alinear con la altura visual del glifo del nombre).
 * El ancho se calcula automáticamente preservando la proporción 4:3.
 */
@Composable
fun CountryFlag(
    countryCode: String?,
    modifier: Modifier = Modifier,
    height: Dp = 15.dp,
    countryName: String? = null,
) {
    if (countryCode.isNullOrBlank()) return
    val code = countryCode.lowercase()
    val context = LocalContext.current

    val request = remember(code) {
        val key = "flag_$code"
        ImageRequest.Builder(context)
            .data("file:///android_asset/flags/$code.svg")
            .memoryCacheKey(key)
            .placeholderMemoryCacheKey(key)
            .build()
    }

    AsyncImage(
        model = request,
        contentDescription = countryName,
        contentScale = ContentScale.Crop,
        modifier = modifier
            .size(width = height * 4 / 3, height = height)
            .clip(RoundedCornerShape(2.dp)),
    )
}
